### ISSUE TITLE
Fix S2589 FP: When local is assigned in for loop

### ISSUE THREAD

**Issue by drieseng** [SonarSource/sonar-dotnet#7088]
### Description

When a local variable is initialized in a for loop, then **S2589** reports a  false positive if you check whether the local variable has a value.

### Repro steps

Compile the following code:
```c#
#nullable enable

using System;
using System.Text;

public sealed class Program
{
    private static void Main(string[] args)
    {
        int? previous = null;
        StringBuilder? sb = null;

        for (var i = 0; i < 5; i++)
        {
            if (previous == null)
            {
                previous = 667;
                continue;
            }

            if (sb is null)
            {
                sb = new StringBuilder();
            }

            sb.Append(i);
        }

        if (sb is null)
        {
            Console.WriteLine("NULL");
        }
    }
}
```

### Expected behavior

S2589 is not reported.

### Actual behavior

S2589 is reported for both line 21 and 29.

### Related information

* SonarAnalyzer.CSharp version: 8.55.0.65544
* Visual Studio version: 17.5.3

**Comment by Tim-Pohlmann** [comment]
Hello, I confirm this as an FP. Thank you for the report!

**Comment by Tim-Pohlmann** [comment]
For future reference: To fix this, our SE engine would need to iterate the loop at least three times. Right now it is only doing it twice.

**Comment by Norvand** [comment]
@Tim-Pohlmann  I have a similar issue.  Is this the same case?
```cs
static void Main(string[] args)
{
    int first = 0;
    int last;

    int row = 0;
    while (true)
    {
        if (row > 5 && first == 0) // first == 0: Warning S2589
        {
            first = row;
        }

        if (row > 10)
        {
            last = row;
            break;
        }

        row++;
    }

    Console.WriteLine($"{first} : {last}");
    Console.ReadKey();
}
```
### Related information
* SonarAnalyzer.CSharp version: 9.20.0.85982
* Visual Studio version: 17.9.0
* .Net Framework 4.7

**Comment by pavel-mikula-sonarsource** [comment]
@Norvand Yes, it looks like the same root cause
